src/widget/wtrackmenu.h

```diff
@@ -0,0 +1,234 @@
+#pragma once
+
+#include <QMenu>
+
+#include "library/dao/playlistdao.h"
+#include "library/dlgtagfetcher.h"
+#include "library/dlgtrackinfo.h"
+#include "library/trackmodel.h"
+
+class ControlProxy;
+class DlgTagFetcher;
+class DlgTrackInfo;
+class ExternalTrackCollection;
+class QAction;
+class QWidget;
+class TrackCollectionManager;
+class WColorPickerAction;
+class WCoverArtMenu;
+
+class WTrackMenu : public QMenu {
+    Q_OBJECT
+  public:
+    enum Feature {
+        AutoDJ = 1,
+        // The loadTrackToPlayer signal emitted from this class must be handled to make LoadTo work.
+        LoadTo = 1 << 1,
+        Playlist = 1 << 2,
+        Crate = 1 << 3,
+        Remove = 1 << 4,
+        Metadata = 1 << 5,
+        Reset = 1 << 6,
+        BPM = 1 << 7,
+        Color = 1 << 8,
+        HideUnhidePurge = 1 << 9,
+        FileBrowser = 1 << 10,
+        Properties = 1 << 11,
+        TrackModelFeatures = Remove | HideUnhidePurge,
+        All = AutoDJ | LoadTo | Playlist | Crate | Remove | Metadata | Reset |
+                BPM | Color | HideUnhidePurge | FileBrowser | Properties
+    };
+    Q_DECLARE_FLAGS(Features, Feature)
+
+    WTrackMenu(QWidget* parent,
+            UserSettingsPointer pConfig,
+            TrackCollectionManager* pTrackCollectionManager,
+            Features flags = Feature::All,
+            TrackModel* trackModel = nullptr);
+
+    ~WTrackMenu() {
+    }
+
+    void loadTrack(TrackId trackId);
+    void loadTrack(QModelIndex index);
+    void loadTracks(TrackIdList trackList);
+    void loadTracks(QModelIndexList indexList);
+
+    // Override default popup method from class QMenu
+    void popup(const QPoint& pos, QAction* at = nullptr);
```

Comment:
add missing `override`

Reply:
`QMenu::popup(const QPoint& pos, QAction* at = nullptr)` is not a virtual function. So this redefinition is hiding the parent function.
This function only adds a check to popup menu only when at least one track is loaded in `WTrackMenu`. Although this might not be a good design, this makes the usage very clean.

Suggested changes:
1. Rename this function to avoid hiding. But this adds overhead on the client class to use non-conventional function to popup the menu.
2. Expect the client class to check empty list on its own before popup.
3. Entirely remove the need for this function and call popup internally from `WTrackMenu` after `loadTracks` is called.

Reply:
> Rename this function to avoid hiding. But this adds overhead on the client class to use non-conventional function to popup the menu.

I would suggest this, but because WTrackMenu subclasses QMenu, someone writing code using WTrackMenu may look at the QMenu documentation and decide to call `popup`.

> Expect the client class to check empty list on its own before popup.

This would leak the abstraction of what tracks are loaded into WTrackMenu.

> Entirely remove the need for this function and call popup internally from WTrackMenu after loadTracks is called.

Hmm, this might be okay if we rename `loadTracks` to `loadTracksAndPopup`... but that still leaves the problem that someone might still call QMenu::popup.

I suggest leaving it how it is but adding a comment noting this hides a non-virtual method of QMenu. @uklotzde what do you think?

Reply:
Ping @uklotzde, I think this is the last remaining issue to take care of in this PR.

Reply:
Please leave a warning that we are overwriting and hiding the non-virtual function QMenu::popup() here on purpose.